From 592d5bd932a0d22739b22432a6ee9597b6934a97 Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Mon, 8 Sep 2014 14:30:52 -0400
Subject: [scsi] hpsa: do not attempt to flush the cache on locked up controllers

Message-id: <1410186734-19986-2-git-send-email-thenzl@redhat.com>
Patchwork-id: 92022
O-Subject: [RHEL6.6 e-stor PATCH 01/83] hpsa: do not attempt to flush the cache on locked up controllers
Bugzilla: 1069185
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

There's no point in trying since it can't work, and if you do
try, it will just hang the system on shutdown.

Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
(cherry picked from commit 702890e3b910cd14c3d0ee3b4842d7e946e66b8d)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index 403b9a7..11c05c2 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -4892,6 +4892,15 @@ static void hpsa_flush_cache(struct ctlr_info *h)
 {
  char *flush_buf;
  struct CommandList *c;
+ unsigned long flags;
+
+ /* Don't bother trying to flush the cache if locked up */
+ spin_lock_irqsave(&h->lock, flags);
+ if (unlikely(h->lockup_detected)) {
+  spin_unlock_irqrestore(&h->lock, flags);
+  return;
+ }
+ spin_unlock_irqrestore(&h->lock, flags);
 
  flush_buf = kzalloc(4, GFP_KERNEL);
  if (!flush_buf)
-- 
1.7.1